src/transformers/models/clvp/modeling_clvp.py

```diff
@@ -0,0 +1,1873 @@
+    CLVPTextConfig,
+    PretrainedConfig,
+)
+
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "susnato/clvp_dev"
+
+CLVP_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "susnato/clvp_dev",
+    # See all CLVP models at https://huggingface.co/models?filter=clvp
+]
+
+
+# Copied from transformers.models.bart.modeling_bart._make_causal_mask
+def _make_causal_mask(
+    input_ids_shape: torch.Size, dtype: torch.dtype, device: torch.device, past_key_values_length: int = 0
+):
+    """
+    Make causal mask used for bi-directional self-attention.
+    """
+    bsz, tgt_len = input_ids_shape
+    mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min, device=device)
+    mask_cond = torch.arange(mask.size(-1), device=device)
+    mask.masked_fill_(mask_cond < (mask_cond + 1).view(mask.size(-1), 1), 0)
+    mask = mask.to(dtype)
+
+    if past_key_values_length > 0:
+        mask = torch.cat([torch.zeros(tgt_len, past_key_values_length, dtype=dtype, device=device), mask], dim=-1)
+    return mask[None, None, :, :].expand(bsz, 1, tgt_len, tgt_len + past_key_values_length)
+
+
+# Copied from transformers.models.bart.modeling_bart._expand_mask
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    bsz, src_len = mask.size()
+    tgt_len = tgt_len if tgt_len is not None else src_len
+
+    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
+
+    inverted_mask = 1.0 - expanded_mask
+
+    return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
+
+
+# Copied from transformers.models.clip.modeling_clip.contrastive_loss
+def contrastive_loss(logits: torch.Tensor) -> torch.Tensor:
+    return nn.functional.cross_entropy(logits, torch.arange(len(logits), device=logits.device))
+
+
+# Copied from transformers.models.clip.modeling_clip.clip_loss with clip->clvp, image_loss->speech_loss
+def clvp_loss(similarity: torch.Tensor) -> torch.Tensor:
+    caption_loss = contrastive_loss(similarity)
+    speech_loss = contrastive_loss(similarity.t())
+    return (caption_loss + speech_loss) / 2.0
+
+
+def rotate_half(state):
+    """
+    This method splits the state into two parts and then rotates the second part by 90 degrees.
+    """
+    state_shape = state.size()
+    state = state.view([*state_shape[:-1]] + [2, state_shape[-1] // 2])
+    state_part_1, state_part_2 = state.unbind(dim=-2)
+    return torch.cat((-state_part_2, state_part_1), dim=-1)
+
+
+def apply_rotary_pos_emb(state, freqs):
+    """
+    Applies rotary position embeddings on state.
+    """
+    pos_emb_len = freqs.shape[-1]
+    state_l, state_r = state[..., :pos_emb_len], state[..., pos_emb_len:]
+
+    freqs = freqs[:, :, -state_l.shape[-1] :]
+    state_l = (state_l * freqs.cos()) + (rotate_half(state_l) * freqs.sin())
+
+    return torch.cat([state_l, state_r], dim=-1)
```

Comment:
let's try to use the rotary embedding from Llama 😉

Reply:
we can use the method for `rotate_half` from llama but for `apply_rotary_pos_emb` we have some additional logic where we apply the embedding to the left portion of the tensor, so must keep `apply_rotary_pos_emb` as it is.

Reply:
`Persimmon` also used a logic where it is only applied on half of the input! 
Anyway, the simpler the better!